### DISCUSSION TITLE
WASM Throttle Block Implementation

### DISCUSSION THREAD

**Discussion by JuliDi** [FutureSDR/FutureSDR#134] [Q&A] [answered]
I've been trying to make the throttle block a wasm-compatible. However, that was not successful yet.
Using [web-time](https://crates.io/crates/web-time) it is possible to replace the uses of `std::time::Instant`, but I could not figure out how to do the sleep. 

There is already a wasm-compatible async sleep used in the flowgraph here https://github.com/FutureSDR/FutureSDR/blob/faef89be2b9a94fbc468a7d207dddb14a3a59f59/src/runtime/flowgraph.rs#L217 but that Future does not implement `Send` and can therefore not be used in here https://github.com/FutureSDR/FutureSDR/blob/faef89be2b9a94fbc468a7d207dddb14a3a59f59/src/blocks/throttle.rs#L98

Any ideas how to make this work?

I am trying to just get some artificially generated data out of the flowgraph in wasm, but data seems to never reach the sink unless throttled. It does work when I compile it for linux.

Once I get that running I would like to also make the websocket sink wasm compatible using [tokio-tungstenite-wasm](https://crates.io/crates/tokio-tungstenite-wasm). Working on a small demo that uses the slint GUI framework for plotting FutureSDR data. It works quite nicely when running slint in the browser and FutureSDR streaming the data from the "desktop" binary via websockets. But I'd like to have it self-contained in a single wasm application.

**Comment by bastibl** [comment, marked as answer]
I tried to add it but have no real test case at the moment, could you checkout the most recent commit?

At the moment, the wasm stuff is all single-threaded (i.e., non-Send) but can be spawned in a WebWorker. The `block_on` function was missing the compile switch, which should be added now. Extending also the wasm stuff to multi-threaded would be possible but that's quite a bit of work and one needs to dive in to shared array buffers.

**Reply by JuliDi** [reply]
Thank you, I will give it a try and report back.

**Reply by JuliDi** [reply]
I think what is still missing is a wasm compatible implementation of time. Now I'm getting `time not implemented on this platform`, presumably due to `std::time::Instant` being used. That one I was able to fix yesterday, so I can just rebase that branch and make a PR. Let's see how that works.

**Comment by bastibl** [comment]
Regarding your issue that something doesn't reach a block, that's often related to buffer history. A native block uses the circular buffer, where this is no issue. On wasm we use the slab buffer, which might need manual history/overlap configuration. See for example: https://github.com/FutureSDR/FutureSDR/blob/main/examples/zigbee/src/wasm_worker.rs#L100

**Reply by JuliDi** [reply]
What do I need to configure there and where does the buffer go? Without it, after a couple of seconds the "sink start" is logged and immediately after I get the "sink end", and then it starts plotting, essentially blocking the event loop, I guess. What I would like to see is that every couple of milliseconds the sink gets some data, sends it for plotting and then the same starts again. I've got that running with websockets, i.e., essentially the websocket example from FutureSDR and then instead of the flowgraph running I have a websocket receiver that pushes the data to the GUI.

This is the code I use to setup the flowgraph and I honestly have no clue between which blocks something like `[Slab::with_config(32768, 2, 128)]`  would go and what settings are correct.

```rust
    let mut fg = Flowgraph::new();

    let src = SignalSourceBuilder::<Complex32>::sin(480.0, 48_000.0).build();
    let noise =
        Apply::new(|i: &Complex32| i + rand::thread_rng().gen_range(-100..100) as f32 / 100.0);

    const FFT_SIZE: usize = 512;
    let fft = Fft::with_options(FFT_SIZE, FftDirection::Forward, true, None);
    let throttle = Throttle::<Complex32>::new(FFT_SIZE as f64);
    let mag = Apply::new(|x: &Complex32| x.norm());


    let snk = Sink::new(move |x: &f32| {
        console::log_1(&"sink start!".into());
        let x = *x;
        // Send it to the GUI for plotting
        console::log_1(&"sink end".into());
    });
    
    connect!(fg, src > throttle > mag > snk);
```

**Reply by bastibl** [reply]
I copied the code into the wasm example and it worked as expected. If you actually use the fft block, you'll have to add a buffer with 512 overlap (last parameter) before the fft, because it need at least 512 items to progress (`connect!(... noise [Slab::with_config(...)] fft > ...`). I'm not sure, I get the point with the sink start/end. Maybe because I don't fully understand the GUI integration. But notice that the sink is called for every sample, so it might not be ideal to get samples into the GUI. You could maybe have a look at the spectrum example or share the complete project code, if this is possible.
https://www.youtube.com/watch?v=YcV8t-0kfGA

**Reply by JuliDi** [reply]
Very interesting, because I can still reproduce this behavior and cannot quite tell the difference to the spectrum example. It looks like the FutureSDR runtime computes a lot of samples at once and then they enter the sink at a rate that blocks the eventloop.

I have added you to the repo where that code lives, if you want to have a look. It's on the `futuresdr-wasm` branch and you can run the example with `SLINT_STYLE=fluent-light  wasm-pack build  --out-dir pkg  --dev --target web --features webdemo`. I tend to use `simple-http-server` as webserver, but I think any other works fine.

Might be that the way the Slint event loop polls the runtime is causing trouble, but I've managed to just "replay" a vector, i.e., iterate over chunks of a vector, send that data to the rendering function and sleep for 100 ms or so. And that works just fine and it seems like the Slint eventloop polls that one often enough (that code is on the main branch, for reference).

After a few seconds of nothing happening on the GUI, the logs start looking like this: 
![image](https://github.com/FutureSDR/FutureSDR/assets/20155974/7694bf80-6178-4a35-badd-568707373ca2)
So I guess that the SDR runtime blocks the GUI eventloop inside the sink and once it unblocks, all the rendering tasks are run, freezing the GUI.

The code is pretty much this:
```rust
#[cfg_attr(target_arch = "wasm32", wasm_bindgen(start))]
pub fn main() {
    // This provides better error messages in debug mode.
    // It's disabled in release mode so it doesn't bloat up the file size.
    #[cfg(all(debug_assertions, target_arch = "wasm32"))]
    console_error_panic_hook::set_once();

    let app = MainWindow::new().unwrap();

    let window_weak = app.as_weak();

    slint::spawn_local(setup_flowgraph(window_weak)).unwrap();

    console::log_1(&format!("run app").into());
    app.run().unwrap();
}

async fn setup_flowgraph(window: Weak<MainWindow>) -> Result<()> {
    let window = window.clone();

    let mut fg = Flowgraph::new();

    let src = SignalSourceBuilder::<Complex32>::sin(480.0, 48_000.0).build();
    let noise =
        Apply::new(|i: &Complex32| i + rand::thread_rng().gen_range(-100..100) as f32 / 100.0);

    const FFT_SIZE: usize = 512;
    let fft = Fft::with_options(FFT_SIZE, FftDirection::Forward, true, None);
    let throttle = Throttle::<Complex32>::new(FFT_SIZE as f64);
    let mag = Apply::new(|x: &Complex32| x.norm());


    let snk = Sink::new(move |x: &f32| {
        console::log_1(&"sink start!".into());
        let x = *x;
        // Send it to the GUI for plotting
        window
            .upgrade_in_event_loop(move |app| {
                app.set_plot_frame(render_plot(vec![x, 20.0, 50.0, 60.0], &app))
            })
            .unwrap();
        console::log_1(&"sink end".into());
    });

    connect!(fg, src > throttle > mag > snk);

    //connect!(fg, src  > noise [Slab::with_config(FFT_SIZE*8,2,FFT_SIZE)] fft  > throttle > mag  > snk);

    console::log_1(&format!("run async fg").into());
    Runtime::new().run_async(fg).await?;

    Ok(())
}
```

**Reply by bastibl** [reply]
The Slab buffer has a default size of 32k. Filling it with 8-byte complex values at a rate of 48k should give you a full buffer every 85ms. The buffer is then passed in one chunk to the subsequent blocks, which will process the whole buffer in their call to work.

I guess your flowgraph is running way slower, because the sink doesn't forward the whole buffer of samples to the GUI but one after the other. Logging is also horribly slow. And logging twice per sample will slow things down to a level where things are no longer usable. Also try compiling in release mode. For wasm this makes a huge difference. The debug build are basically unusable.

I'd use a sink that forwards either a fixed-size vector through a websocket
https://github.com/FutureSDR/FutureSDR/blob/main/examples/spectrum/src/bin/cpu.rs#L24-L26

or forward them through a channel:
https://github.com/FutureSDR/FutureSDR/blob/main/src/blocks/channel_sink.rs

If this doesn't help, I can have a look at the project.

**Reply by JuliDi** [reply]
Alright, that is definitely something I was not aware of. So I would essentially have to place a 1-sample buffer everywhere if I want to get one sample at a time?
Like
```rust
connect!(fg, src > throttle  [Slab::with_config(2,1,1)] mag  [Slab::with_config(2,1,1)] snk);
```
and for the FFT then something like
```rust
connect!(fg, src  > noise [Slab::with_config(FFT_SIZE*8,2,FFT_SIZE)] fft  [Slab::with_config(FFT_SIZE*8,2,FFT_SIZE)] throttle [Slab::with_config(FFT_SIZE*8,2,FFT_SIZE)] mag  [Slab::with_config(FFT_SIZE*8,2,FFT_SIZE)] snk);
```
I don't yet understand how the Slab buffer is configured properly, https://docs.rs/futuresdr/latest/futuresdr/runtime/buffer/slab/struct.Slab.html#method.with_config does not describe what `reserved_items` does and how this results in the "overlap" necessary for the FFT.

I agree that logging etc slows things down and the rate in the above code is now too high, but it hasn't been working with the throttle block set to 1 sample. So presumably it's due to the buffer being filled because 2 log messages and drawing like 4 samples is not going to take more than a second, even on debug build (it's the same in release, either).

The websocket idea is already implemented and working, but then it's not entirely self-contained in a web app (websocket TX running on the machine locally).
The channel sink looks more like what I need, so I could just have another task awaiting the RX part of the channel and pass it to the rendering function. With the channel sink, is there a way to send a fixed-size vector, as well? Ideally I want the same behavior as the websocket with a fixed-size output. From what I see in the code, that depends on the buffer, or not?

Thanks for your help, I really appreciate that you're taking the time!

**Reply by bastibl** [reply]
Not sure why you want to have 1 sample at a time. With a sample rate of 48k and 5 blocks in a chain this would force 48k * 5 work() calls per second for no reason. This will probably cause problems, i.e., saturate the rendering thread in a debug build. If you really want to process each sample individually, there is no real reason to use an sdr runtime, no?

Regarding the slab buffer: this is just some allocated memory. Once it is filled, it is forwarded to the downstream block with reads from the buffer. To avoid stalling the upstream block, while the downstream block reads, you can have more than one buffer per edge. With the default of 2, it is just double buffering, i.e., upstream can fill one buffer, while downstream reads from the other. This is the second parameter. The first one is the buffer size in bytes. That means if you really want one sample per buffer, you'd have to configure 4 or 8 in your case, depending on the edge (if it is float or complex float).

The last parameter is for the overlap/history. Assume you have an fft block, which needs 512 samples to do something. In the current buffer, there are only 500 samples left. And its a slab buffer not a circular one. Then what to do? You could force the block to store this locally, i.e., make a copy of the 500 samples, consume them, and wait for the next buffer. But this would force non-trivial logic for each block. What we are doing is copy over up to `reserved_items` samples to the next buffer. For this, we leave some space in every buffer. See this blog post for some more details: https://www.futuresdr.org/blog/red-slab/

**Reply by JuliDi** [reply]
The 1 sample at a time was just for testing/debugging purposes until I figured out how to best collect a full vector of FFT output to send to the renderer. However: wouldn't the Throttle block cause this to slow down regardless of the sample rate?

So from what I understand now, I would rather have a lower sample rate (not 48k, more like `2*FFT_SIZE`) and have no throttling here (because the block already outputs only 2k samples per second). Then this is fed to the FFT through a Slab Buffer of size 8*FFT_SIZE (Complex32 is 8 bytes and the buffer should be passed downstream to the FFT Block after FFT_SIZE samples are available), which then produces 2 vectors of FFT_SIZE per second.

And then I could build a modified ChannelSink that implements the Websocket Sink's behavior of transmitting a fixed FFT_SIZE vector, so the renderer gets 2 vectors of size FFT_SIZE per second.

On the other hand, if I understand the slab buffers correctly now, everything _should_ be working when I add a Throttle block, throttling to "Signal Generator Sample Rate" samples per second (e.g., the 2* FFT_SIZE) and combine this with a modified ChannelSink, which should get 2*FFT_SIZE samples per second, if they are available. 

Another thing might be to integrate the WASM websocket, but I couldn't get that working with my websocket listener, but that might be solvable - iirc it's just a different websocket crate that what I had been using with the non-WASM websocket sink.
Edit: The WASM Websocket idea probably won't work because the "normal" Websocket Sink provides a websocket server itself and the WASM Websocket wants to connect to one. And the GUI part, running on WASM, as well, cannot be a server either (afaik the browser cannot act as a server and accept incoming connections, so a channel is the way to go here).

**Reply by bastibl** [reply]
Here is an adapted ChannelSink with fixed size vectors that I use for the egui example:
https://github.com/FutureSDR/FutureSDR/blob/main/examples/egui/src/channel_sink.rs
If I understand you correctly that's exactly what you want. Looks like this:
https://x.com/bastibl/status/1754078443912966567

I don't get your reasoning for the buffer sizes and rates. When you do an an fft, you don't have to enforce that there are only FFT_SIZE samples in the buffer. And the minimal example already worked. So I'm not sure about the issue. When you really want to minimize or enforce buffers, you could set the size to 8*FFT_SIZE and the reserved items to 0. But again that part does not make a lot of sense to me.

**Reply by JuliDi** [reply]
> If you actually use the fft block, you'll have to add a buffer with 512 overlap (last parameter) before the fft, because it need at least 512 items to progress

That's why I wanted to add the buffer. Now I'm confused in which situations I would need a custom buffer and when the default is alright.

The minimal example ignores that the flowgraph is run inside Slint's eventloop, so that might be an issue. However, it does yield samples after being stuck for a few seconds, it's more of the timing and how regularly they arrive at the sink.

I will try to pinpoint where exactly it gets stuck and compare it with the wasm example. Because it does seem to work when I run it on the desktop, so it must be some WASM (+Slint) peculiarity.

**Reply by bastibl** [reply]
Buffers are complicated :-) For your FFT you have two options. You  could either make the number of items that fit into the buffer a multiple of the FFT size (with 0 reserved items). In this case, you'll never come into the situation that only 500 or whatever items are left and you cannot proceed. Or you create the buffer with an arbitrary size but ensure that the copy-over mechanism (see blog post) makes sure that once fewer items than the FFT size are left in the buffer, the items will be copied over to the next buffer, which ensures that the block does not get stuck (i.e., getting called again and again with the remaining items that are not enough to proceed). This is configured with the reserved_items. The middle parameter, I always set to two.

**Reply by JuliDi** [reply]
I see, thanks for the explanation! Now it's more clear what impact the buffer has - and what is related to how the runtime itself works.
And on top of that the influence of the async runtime/WASM remains to be determined... ;)

**Reply by JuliDi** [reply]
Using `connect!(fg, src > throttle [Slab::with_config(4,2,0)] mag [Slab::with_config(4,2,0)] snk);`  does indeed do the trick with my above example when I throttle to 1 sample per second. Now I can continue extending it to the FFT. Thanks!

**Reply by bastibl** [reply]
Arg, in that second I made a PR for your project xD

**Reply by JuliDi** [reply]
Oh now you're stealing all the fun! :D I've havent started working on the FFT part, I only confitmed that it's working with the single-sample sink, but I will check out your PR tomorrow. Thanks a lot!
